### PR TITLE
[consensus] Add failed_proposers to BlockMetadata

### PR DESCRIPTION
**Pull request by ikabiljo**
For: https://github.com/aptos-labs/aptos-core/issues/1395

### CODE REVIEW

File: consensus/consensus-types/src/block.rs [hunk 1]
```diff
@@ -360,6 +366,22 @@ impl Block {
...
+        validators: &[AccountAddress],
+        failed_authors: &[(Round, Author)],
+    ) -> Vec<u32> {
+        failed_authors
+            .iter()
+            .map(|(_round, failed_author)| {
+                validators
+                    .iter()
+                    .position(|&v| v == *failed_author)
+                    .unwrap()
```
Comment (sitalkedia) [inline consensus/consensus-types/src/block.rs:380]: nit: instead of `unwrap()`, let's use `expect()` with a proper error message.

File: types/src/block_metadata.rs [hunk 2]
```diff
@@ -33,6 +33,7 @@ pub struct BlockMetadata {
     round: u64,
     previous_block_votes: Vec<bool>,
     proposer: AccountAddress,
+    failed_proposer_indices: Vec<u32>,
```
Comment (sitalkedia) [inline types/src/block_metadata.rs:36]: Is it possible that a proposer failed multiple times? Do we want to take that into account here or that is an overkill?

Comment (ikabiljo) [reply types/src/block_metadata.rs:36]: Yeah, same proposer can fail multiple times, and will be multiple time repeated in this list.

list is in the rounds order

Comment (ikabiljo) [reply types/src/block_metadata.rs:36]: you can see the example in the recover_on_restart test here :
https://github.com/aptos-labs/aptos-core/pull/1481/commits/b5416f40620af8016e1ce72bb686d9917c2f07be#diff-e635da0173fb15af50f16e92fbeb662e4da79a18ea68cd522e2c32e463cec5c4R743

**Review by sitalkedia — APPROVED** [review]
LGTM

File: consensus/consensus-types/src/block.rs [hunk 3]
```diff
@@ -361,6 +367,27 @@ impl Block {
...
+    fn failed_authors_to_indices(
+        validators: &[AccountAddress],
+        failed_authors: &[(Round, Author)],
+    ) -> Vec<u32> {
+        failed_authors
+            .iter()
+            .map(|(_round, failed_author)| {
+                validators
+                    .iter()
+                    .position(|&v| v == *failed_author)
```
Comment (movekevin) [inline consensus/consensus-types/src/block.rs:380]: Since you're newly adding this, can you use the validator index from validator config instead? https://github.com/aptos-labs/aptos-core/blob/main/types/src/validator_config.rs#L42

This can ensure that in future when we deprecate validator set sorting (expensive to do on chain), we can still ensure that this works correctly

File: aptos-move/framework/aptos-framework/sources/Block.move [hunk 4]
```diff
@@ -84,6 +86,7 @@ module AptosFramework::Block {
                 round,
                 previous_block_votes,
                 proposer,
+                failed_proposer_indices,
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/Block.move:89]: How do we ensure that other validators can validate failed_proposer_indices is correct and not maliciously altered by the proposer? Is this done in a separate PR?

**Comment by movekevin** [comment]
I think it'd be great if you can add a PR description with more details on:

How (do they work) and why the code changes?
What is the overall high level design (a short design doc would be great as well)?
What are the other steps that need to be done for proposer reputation to work from end to end? It's hard to understand how this particular change fits into the bigger picture.

**Review by davidiw — COMMENTED** [review]
besides Kevin's comments, looks good
we'll really need to start benchmarking operations that occur per block and at epochs.

**Comment by ikabiljo** [comment]
@movekevin , @davidiw - attached issue is for solving the improved leader selection, and has all the relevant pull requests attached - #1395

Most specifically, https://github.com/aptos-labs/aptos-core/pull/1523 has the information about the new leader selection logic.

I've added the summary of PRs into the parent issue. Let me know if you want me to add more details to the parent issue, or to that other pull request, but I feel like this PR itself is very clear on what it does by itself.

File: consensus/consensus-types/src/block.rs (hunk 3 above)
Comment (movekevin) [reply consensus/consensus-types/src/block.rs:380]: Turns out this is not easy to do as block execution doesn't readily have access to the validator set/validator info needed to fetch validator indices. We can do this later in a different PR when we clean up the sorting logic.
